// @minecraft/server/index.js
export const mockState = {
    intervals: [],
    players: [],
    messagesSent: [],
    titlesSet: [],
    subtitlesSet: [],
    soundsPlayed: [],
    reset() {
        this.intervals = [];
        this.players = [];
        this.messagesSent = [];
        this.titlesSet = [];
        this.subtitlesSet = [];
        this.soundsPlayed = [];
    }
};

export const system = {
    runInterval(callback, ticks) {
        mockState.intervals.push({ callback, ticks });
        return mockState.intervals.length;
    }
};

class MockPlayer {
    constructor(name) {
        this.name = name;
        this.dynamicProperties = {};
        this.messagesReceived = [];
        this.dimension = {
            spawnItem: (itemStack, location) => {
                mockState.spawnedItems = mockState.spawnedItems || [];
                mockState.spawnedItems.push({ itemStack, location, player: this.name });
            }
        };
        this.location = { x: 0, y: 0, z: 0 };
        this.inventory = {
            container: {
                addItem: (itemStack) => {
                    mockState.addedItems = mockState.addedItems || [];
                    mockState.addedItems.push({ itemStack, player: this.name });
                    return null; // Return null to simulate all items added successfully
                }
            }
        };
        this.onScreenDisplay = {
            setTitle: (title, options) => {
                mockState.titlesSet.push({ player: this.name, title });
                if (options) {
                    if (options.subtitle) {
                        mockState.subtitlesSet.push({ player: this.name, subtitle: options.subtitle });
                    }
                    if (options.stayDuration === undefined || options.stayDuration === null || typeof options.stayDuration !== 'number') {
                        throw new TypeError("Native setter cannot be assigned null or undefined. Interface property ['stayDuration'] expected type: number");
                    }
                }
            }
        };
    }

    getDynamicProperty(key) {
        return this.dynamicProperties[key];
    }

    setDynamicProperty(key, value) {
        this.dynamicProperties[key] = value;
    }

    getComponent(componentId) {
        if (componentId === 'inventory') {
            return this.inventory;
        }
        return null;
    }

    sendMessage(message) {
        this.messagesReceived.push(message);
        mockState.messagesSent.push({ player: this.name, message });
    }

    playSound(soundId, options) {
        mockState.soundsPlayed.push({ player: this.name, soundId, options });
    }
}

export const world = {
    getAllPlayers() {
        return mockState.players;
    },
    beforeEvents: {
        playerLeave: {
            subscribe(callback) {
                mockState.playerLeaveSubscribe = callback;
            }
        },
        itemUseOn: {
            subscribe(callback) {
                mockState.itemUseOnSubscribe = callback;
            }
        },
        worldInitialize: {
            subscribe(callback) {
                mockState.worldInitializeSubscribe = callback;
                // Automatically run callback to register dynamic properties during import/setup
                const mockEvent = {
                    registerPlayerDynamicProperties(def) {
                        mockState.registeredPlayerDynamicProperties = mockState.registeredPlayerDynamicProperties || [];
                        mockState.registeredPlayerDynamicProperties.push(def);
                    }
                };
                callback(mockEvent);
            }
        }
    },
    afterEvents: {
        playerSpawn: {
            subscribe(callback) {
                mockState.playerSpawnSubscribe = callback;
            }
        },
        playerBreakBlock: {
            subscribe(callback) {
                mockState.playerBreakBlockSubscribe = callback;
            }
        },
        entityDie: {
            subscribe(callback) {
                mockState.entityDieSubscribe = callback;
            }
        }
    }
};

export class DynamicPropertiesDefinition {
    constructor() {
        this.strings = {};
    }
    defineString(key, length) {
        this.strings[key] = length;
    }
}

export class ItemStack {
    constructor(typeId, amount) {
        this.typeId = typeId;
        this.amount = amount;
    }
}

export function addMockPlayer(name) {
    const p = new MockPlayer(name);
    mockState.players.push(p);
    return p;
}
