// @minecraft/server/index.js
export const mockState = {
    intervals: [],
    players: [],
    messagesSent: [],
    titlesSet: [],
    subtitlesSet: [],
    soundsPlayed: [],
    reset() {
        this.intervals = [];
        this.players = [];
        this.messagesSent = [];
        this.titlesSet = [];
        this.subtitlesSet = [];
        this.soundsPlayed = [];
    }
};

export const system = {
    runInterval(callback, ticks) {
        mockState.intervals.push({ callback, ticks });
        return mockState.intervals.length;
    }
};

class MockPlayer {
    constructor(name) {
        this.name = name;
        this.onScreenDisplay = {
            setTitle: (title, options) => {
                mockState.titlesSet.push({ player: this.name, title });
                if (options && options.subtitle) {
                    mockState.subtitlesSet.push({ player: this.name, subtitle: options.subtitle });
                }
            }
        };
    }

    sendMessage(message) {
        mockState.messagesSent.push({ player: this.name, message });
    }

    playSound(soundId, options) {
        mockState.soundsPlayed.push({ player: this.name, soundId, options });
    }
}

export const world = {
    getAllPlayers() {
        return mockState.players;
    }
};

export function addMockPlayer(name) {
    const p = new MockPlayer(name);
    mockState.players.push(p);
    return p;
}
